pkg/container-disk/container-disk_test.go

```diff
@@ -503,7 +505,7 @@ func createMigrationSourcePod(vmi *v1.VirtualMachineInstance) *k8sv1.Pod {
 		status := k8sv1.ContainerStatus{
 			Name:    container.Name,
 			Image:   container.Image,
-			ImageID: fmt.Sprintf("finalimg@sha256:%v", idx),
+			ImageID: fmt.Sprintf("finalimg@sha256:%v", strings.Repeat(strconv.Itoa(idx%10), 64)),
```

Comment:
Why is this necessary?

Reply:
Because now in the regex we have (as expected for a sha256 entry) `{64}` so a shorter string like `1` will not be matched.

Reply:
Thanks, while you are right, we do not validate the digest as we should. e.g sha256 should not use upper cases. 
I would recommend dropping this length restriction now as it is not required for the fix. We can have a look at proper digest validation later.

Note: Also it is OCI recommendation to allow passing unrecognized digests.

Reply:
this is now consistent with test/e2e_node/image_id_test.go from kubevirt/kubevirt,
see: https://github.com/kubernetes/kubernetes/pull/121531/files#diff-936fcc23cf627b77339d016c5b93692ddfea2eb7bab0d98950612643d5c04becR67

Reply:
While that is true it is only test, not a requirement by Kubernetes or production code. Any other format would work with Kubernetes but not with Kubevirt. We also need to backport this to a stable branch and it is unnecessary if not dangerous to backport it.  Please drop it for now and feel free to raise PR with it alone.

Reply:
OK, but then what should we match every random string of any length? every sequence of hex digits of any length?
This was working before because we were mandating the presence of the `sha256:` and we were assuming that anything after that could only be a well shaped digest.
But if we cannot rely anymore on the presence of the `sha256:` prefix, we can only more careful try to really match a legit digest with a well shaped regex.

For instance, we already have a test here named `should fail if it can't detect a reproducible imageID` that is feeding it with:
`pod.Status.ContainerStatuses[0].ImageID = "rubbish"`
in the original code this was working just because `rubbish` doesn't contain `sha256:` but if we remove that anchor it will start failing.

`(sha256:)?([a-zA-Z0-9]+)` will indeed match `rubbish` or even the `docker` prefix in `docker://sha256:4cbffcbac421e150ae84b0973863609494aad70f0a80eaeb64bddd8d92465812`.

Then we can somehow make it a bit more robust correctly matching only hex digits with `[[:xdigit:]]` instead of `[a-zA-Z0-9]`, but in that case something like `(sha256:)?([[:xdigit:]]+)` will also match `d`, `c` and `e` in `docker://sha256:4cbffcbac421e150ae84b0973863609494aad70f0a80eaeb64bddd8d92465812` or `bb` in `rubbish`.

Reply:
Maybe the context was not so clean,
the only usage of this regex is in:
https://github.com/kubevirt/kubevirt/blob/5fe600a4d3b57623744b25ba0d3923782558ce2b/pkg/container-disk/container-disk.go#L435-L448

that should return an error and make
https://github.com/kubevirt/kubevirt/blob/5fe600a4d3b57623744b25ba0d3923782558ce2b/pkg/virt-controller/services/template.go#L263-L266
fail preventing the migration when we are not able to correctly identify a reproducible digest (please read till the end!).

Now, if we start matching random strings here with a weaker regex (because we cannot rely anymore on the presence of the `sha256:` prefix before the digest) we will allow migrations when we shouldn't introducing a bug.

And, by the way, I'm even not 100% sure that the Kubevirt code is correct because
https://github.com/kubevirt/kubevirt/blob/5fe600a4d3b57623744b25ba0d3923782558ce2b/pkg/container-disk/container-disk.go#L447 is assuming that `digestMatches[1]` is the digest of the image while instead it's now (?) the ImageID and not really what the user specified.
See also:
https://github.com/kubernetes/cri-api/commit/aa73fdd38be83e35419eab9bb62f2e8f72834cc7

But maybe we can open another issue and keep this for another PR.

Reply:
But yes, double thinking at this, images are pullable by digest, not by imageID so this is not the right fix and this is even a more serious issue.

For instance now I see:
```
                "containerStatuses": [
                    {
                        "containerID": "cri-o://0f6c7a363011a0de9053a3c7987fa17924c01d77cbf3e40ceb3d5c12c67b3144",
                        "image": "quay.io/kubevirt/virt-launcher@sha256:b9b04d9d06ada4486e6acaa08e99aaa6a050e1843624ba4f291f8d167ebec36b",
                        "imageID": "3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb",
```
where `b9b04d9d06ada4486e6acaa08e99aaa6a050e1843624ba4f291f8d167ebec36b` is the digest and `3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb`

as in:
```bash
$ podman pull quay.io/kubevirt/virt-launcher@sha256:b9b04d9d06ada4486e6acaa08e99aaa6a050e1843624ba4f291f8d167ebec36b
Trying to pull quay.io/kubevirt/virt-launcher@sha256:b9b04d9d06ada4486e6acaa08e99aaa6a050e1843624ba4f291f8d167ebec36b...
Getting image source signatures
Copying blob sha256:e42699acf9bd09021aefebfaa0501560a5b92aa407da43c8d572792c347e881b
Copying blob sha256:fab240f0b46941a8bdc3175dd7215f6970b5e5d0c759620167b190412246f97e
Copying config sha256:3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb
Writing manifest to image destination
3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb
```

`3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb` is the last line there but

```
$ podman pull quay.io/kubevirt/virt-launcher@sha256:3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb
Trying to pull quay.io/kubevirt/virt-launcher@sha256:3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb...
Error: initializing source docker://quay.io/kubevirt/virt-launcher@sha256:3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb: reading manifest sha256:3682862e41fc27b1ac1054a6778dc0b0e24b00bff4b968b83e16573381e853bb in quay.io/kubevirt/virt-launcher: manifest unknown
```
(where we replaced the digest with the imageID as in the KubeVirt code) is not working and so we have even a serious issue here.

Reply:
Thanks, @tiraboschi, I agree this was based on a not-so-strong assumption from the beginning. Perfection would be to ask Kubernetes to provide a digest but that is not a solution in the short term here. The only thing I see as a possible solution is: to check if user user-specified image can be used if not validate imageID. The only second solution I could think of is to revert the crio change, but I am not sure if that will be accepted.